### ISSUE TITLE
Autograd inside hooks is disabled by default?

### ISSUE THREAD

**Issue by tsavchyn** [pytorch/pytorch#17315]
## 🐛 Bug

I was playing around  with the  `autograd` and "intermediate gradients". And discovered strange behaviour (as for me).  Looks like by default `autograd` is disabled inside hooks (using `register_hook` to extract intermediate gradients). But why? To avoid a kind of "backward inside backward"?

### Plain hook works fine.

```python
import torch


def extract(V):
    def hook(grad):
        V.grad = grad
    return hook

X = torch.ones(5, requires_grad=True)
V = X.pow(3)
Y = V.sum()

V.register_hook(extract(V))
Y.backward()

print('X.grad:', X.grad)
print('V.grad:', V.grad)
```
Output:
```
X.grad: tensor([3., 3., 3., 3., 3.])
V.grad: tensor([1., 1., 1., 1., 1.])
```

### Autograd inside hook fails.

```python
import torch


def extract_and_grad(V):
    def hook(grad):
        V.grad = grad
        V.grad.requires_grad_(True)
        print('V.grad:', V.grad)
        U = V.grad.pow(2).sum()
        print('U:', U)
        try:
            dU_dV_grad = torch.autograd.grad(U, V.grad)
            print('V.grad.grad:', dU_dV_grad)
        except RuntimeError as err:
            print(err)
    return hook

X = torch.ones(5, requires_grad=True)
V = X.pow(3)
Y = V.sum()

V.register_hook(extract_and_grad(V))
Y.backward()
```
Output:
```
V.grad: tensor([1., 1., 1., 1., 1.], requires_grad=True)
U: tensor(5.)
element 0 of tensors does not require grad and does not have a grad_fn
```
Ok, it fails because  there is no `grad_fn` for `U`. Looks like we need to turn on `autograd`.

### Autograd inside hook with `torch.enable_grad()` seems to work fine.

```python
import torch


def extract_and_grad_v2(V):
    def hook(grad):
        V.grad = grad
        V.grad.requires_grad_(True)
        print('V.grad:', V.grad)
        
        with torch.enable_grad():
            U = V.grad.pow(2).sum()
            print('U:', U)
        try:
            dU_dV_grad, = torch.autograd.grad(U, V.grad)
            print('V.grad.grad:', dU_dV_grad)
        except RuntimeError as err:
            print(err)
    return hook

X = torch.ones(5, requires_grad=True)
V = X.pow(3)
Y = V.sum()

V.register_hook(extract_and_grad_v2(V))
Y.backward()
```
Output:
```
V.grad: tensor([1., 1., 1., 1., 1.], requires_grad=True)
U: tensor(5., grad_fn=<SumBackward0>)
V.grad.grad: tensor([2., 2., 2., 2., 2.])
```
Now it woks fine. In addition, outside of hooks with gradients it works as usual.

### Autograd outside of hooks works as expected.

```python
import torch


def extract(V):
    def hook(grad):
        V.grad = grad
    return hook

X = torch.ones(5, requires_grad=True)
V = X.pow(3)
Y = V.sum()

V.register_hook(extract(V))
Y.backward()

V.grad.requires_grad_(True)
print('X.grad:', X.grad)
print('V.grad:', V.grad)
U = V.grad.pow(2).sum()
print('U:', U)
try:
    dU_dV_grad, = torch.autograd.grad(U, V.grad)
    print('V.grad.grad:', dU_dV_grad)
except RuntimeError as err:
    print(err)
```
Output:
```
X.grad: tensor([3., 3., 3., 3., 3.])
V.grad: tensor([1., 1., 1., 1., 1.], requires_grad=True)
U: tensor(5., grad_fn=<SumBackward0>)
V.grad.grad: tensor([2., 2., 2., 2., 2.])
```

## Expected behaviour
I expected `autograd`  to work as usual inside hooks (`registered_hook`).
I am not sure is it a bug or smth. I'm just worrying, why does it behave in such way?

## Environment
```
PyTorch version: 1.0.1.post2
CUDA used to build PyTorch: 10.0.130

OS: Linux Mint 19.1 Tessa
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 970
Nvidia driver version: 410.78
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.0.1.post2
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.1           py3.6_cuda10.0.130_cudnn7.4.2_2    pytorch
[conda] torchvision               0.2.1                      py_2    pytorch
```

**Comment by soumith** [comment]
autograd in general is disabled on the backward pass by default.

If you do want to have it enabled, use `torch.autograd.grad(..., create_graph=True)`. See: https://pytorch.org/docs/stable/autograd.html?highlight=grad#torch.autograd.grad

**Comment by tsavchyn** [comment]
Now I see. Thank you!
